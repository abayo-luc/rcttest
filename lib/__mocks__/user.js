export default {
  userInformation: jest.fn().mockReturnValue({id: 'qwer-37'}),
};
