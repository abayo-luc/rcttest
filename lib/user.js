export async function userInformation() {
  return await null;
}
