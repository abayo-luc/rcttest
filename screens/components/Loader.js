import React, {useContext, useEffect} from 'react';
import {StatusBar, View, Text} from 'react-native';

import {LocalizationContext} from '../../localize';

import {userInformation} from '../../lib/user';

/*
From what I gather, this file gets loaded first somehow and
based on userInformation promise above redirects the user
to the login page or home page.

The LocalizationContext is a hook that is badly commented
and confusing on how it really works. Hoping it can be
mocked somehow since its effect/output dont matter for the tests

Tests wanted:

1. render (done)
2. ensure it redirects to Login page if userInformation resolves null
else go to the home page (failing)
*/

const Loader = ({navigation}) => {
  const {initializeAppLanguage} = useContext(LocalizationContext);

  initializeAppLanguage();

  useEffect(() => {
    (async () => {
      const userInfo = await userInformation();
      navigation.navigate(userInfo ? 'App' : 'Login'); // if there is no user, navigate to login.
    })();
  });

  return (
    <View>
      <StatusBar barStyle="default" />
      <Text>Loading...</Text>
    </View>
  );
};

export default Loader;
