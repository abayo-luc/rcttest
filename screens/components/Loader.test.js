/**
 * @format
 */

import 'react-native';
import React from 'react';
import Loader from './Loader';
import renderer from 'react-test-renderer';
import * as user from '../../lib/user';

const props = {
  navigation: {
    navigate: jest.fn(),
  },
};

let wrapper;

describe('Component: Loader', () => {
  beforeAll(() => {
    // spay on use effect, you can even mock what it will return
    jest.spyOn(React, 'useEffect').mockImplementation((f) => f());
  });
  describe('User is absent', () => {
    beforeEach(async () => {
      // await for component to mount, if you don't, wrapper will be created with
      /// userInformation() function pending, and
      // therefore navigation will not be reached
      wrapper = await renderer.create(<Loader {...props} />);
    });
    afterEach(() => {
      jest.resetAllMocks();
    });
    it('should match snapshot', () => {
      // span shot testing
      expect(wrapper).toMatchSnapshot();
    });
    it('redirects to config page when user is not logged in', async () => {
      expect(props.navigation.navigate).toHaveBeenCalledWith('Login');
    });
  });

  describe('User is present', () => {
    beforeEach(async () => {
      user.userInformation = jest
        .fn()
        .mockReturnValue({id: 'qwerty123', name: 'luc', role: 'admin'});
      wrapper = await renderer.create(<Loader {...props} />);
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it('should navigate to app if user is logged in', () => {
      expect(props.navigation.navigate).toHaveBeenCalledWith('App');
    });
  });
});
