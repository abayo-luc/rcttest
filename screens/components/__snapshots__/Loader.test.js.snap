// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Loader User is absent should match snapshot 1`] = `
<View>
  <Text>
    Loading...
  </Text>
</View>
`;
